ROCANA-4303 Don't use scanf

actgardner opened the pull request:
It turns out `sscanf` is only supplied in glibc 2.7 or higher. Use a custom parsing function to get the capture ID.

`objdump` shows the newest symbol we depend on now is from glibc 2.3:

```
Thu Dec 03 17:53:39 vagrant ~ 
1002 $ objdump -T /vagrant/rocana-agent | fgrep GLIBC
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 printf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memset
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ftell
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 snprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 abort
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fseek
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 exit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __assert_fail
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getaddrinfo
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strncmp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fopen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __libc_start_main
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigfillset
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getpid
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_destroy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 vfprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fputc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 freeaddrinfo
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strlen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_create
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __ctype_b_loc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_getstacksize
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strerror
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtol
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memcpy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memmove
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strchr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fread
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 setenv
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __errno_location
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 asprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strcmp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 calloc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 feof
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fclose
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 random
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strncpy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_init
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stderr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_sigmask
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fwrite
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 realloc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 perror
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 gai_strerror
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtod
0000000000407160      DF *UND*  0000000000000000  GLIBC_2.2.5 malloc
00000000004071f0      DF *UND*  0000000000000000  GLIBC_2.2.5 free
```

Commit:
ROCANA-4303 Don't use scanf

adamlamar commented on cgrok/grokre.c:
```diff
@@ -383,6 +383,22 @@ static void grok_capture_add_predicate(grok_t *grok, int capture_id,
+		if (groupName[i] == 0) {
+			break;
+		}
+		if (! ((groupName[i] >= '0' && groupName[i] <= '9') || (groupName[i] >= 'a' && groupName[i] <= 'f'))) {
```
What about `groupName` characters that are capital letters? Should we allow those?

adamlamar commented on cgrok/grokre.c:
```diff
@@ -383,6 +383,22 @@ static void grok_capture_add_predicate(grok_t *grok, int capture_id,
+			return 0;
+		}
+	}
+	*id = strtol(groupName, NULL, 16);
```
Reading the `strtol` manpage, it will stop on the first invalid character. I'm not sure the above loop is necessary, or did you have something else in mind?

Also, `strtol` sets `errno`, so you should be able to correctly return success and failure. <a href="http://linux.die.net/man/3/strtol">Manpage is here</a> and has a nice example section.

actgardner commented:
@adam-rocana After sleeping on it, I rewrote this to be less redundant. I also made a change to the format string itself, to fix an issue where a PCRE group named `deadbeef` (for example) would be parsed as a huge group number (`dead`), and added a unit test.

adamlamar commented on cgrok/grok.h:
```diff
@@ -125,8 +125,8 @@ extern int g_cap_definition;
-#define CAPTURE_ID_LEN 4
-#define CAPTURE_FORMAT "%04x"
+#define CAPTURE_ID_LEN 6
+#define CAPTURE_FORMAT "0x%04x"
```
I like this.

Does it mean we're limited to 99 captures?

actgardner replied:
We're limited to `16^4` captures, we have 4 hex digits to encode capture number.

edit: to be clear, the first group is `0x0000`, the second is `0x0001`, etc. to `0xffff`.

adamlamar replied:
Oh, duh.  2 != 4, and we're talking about hex. Thanks for answering my silly question.

adamlamar commented on cgrok/grokre.c:
```diff
@@ -383,6 +383,17 @@ static void grok_capture_add_predicate(grok_t *grok, int capture_id,
+		return 0;
+	}
+	*id = strtol(groupName, &endOfNum, 16);
+	return (endOfNum-groupName) == 6;
```
Do you mind checking `errno` here? I can see you're checking the length of the string to ensure it is the length of `CAPTURE_LEN`, but we don't actually know if `strtol` succeeded or not.

adamlamar commented:
LGTM
